File: max_subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        localMax = nums[0]
        globalMax = nums[0]
        print("index:", 0,  "local max:", localMax, "global max:", globalMax)
        
        for i in range(1, len(nums)):
            currentNum = nums[i]
            localMax = max(localMax + currentNum, currentNum)
            globalMax = max(globalMax, localMax)
            print("index:", i, "local max:", localMax, "global max:", globalMax)
        
        return globalMax

if __name__ == "__main__":
    array = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    solution = Solution()
    print("the maximum subarray sum:", solution.maxSubArray(array))


"""
output:
('index:', 0, 'local max:', -2, 'global max:', -2)
('index:', 1, 'local max:', 1, 'global max:', 1)
('index:', 2, 'local max:', -2, 'global max:', 1)
('index:', 3, 'local max:', 4, 'global max:', 4)
('index:', 4, 'local max:', 3, 'global max:', 4)
('index:', 5, 'local max:', 5, 'global max:', 5)
('index:', 6, 'local max:', 6, 'global max:', 6)
('index:', 7, 'local max:', 1, 'global max:', 6)
('index:', 8, 'local max:', 5, 'global max:', 6)
('the maximum subarray sum:', 6)
"""

File: max_subarray.go
package main

import "fmt"

func maxSubArray(nums []int) int {
	localMax := nums[0]
	globalMax := nums[0]
	fmt.Println("index: 0", ",local max:", localMax, ",globalMax:", globalMax)

	for i := 1; i < len(nums); i++ {
		currentNum := nums[i]
		localMax = max(localMax+currentNum, currentNum)
		globalMax = max(globalMax, localMax)
		fmt.Println("index:", i, ",local max:", localMax, ",globalMax:", globalMax)
	}
	return globalMax
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	array := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	fmt.Println("the maximum subarray sum:", maxSubArray(array))
}

/*
output:
index: 0 ,local max: -2 ,globalMax: -2
index: 1 ,local max: 1 ,globalMax: 1
index: 2 ,local max: -2 ,globalMax: 1
index: 3 ,local max: 4 ,globalMax: 4
index: 4 ,local max: 3 ,globalMax: 4
index: 5 ,local max: 5 ,globalMax: 5
index: 6 ,local max: 6 ,globalMax: 6
index: 7 ,local max: 1 ,globalMax: 6
index: 8 ,local max: 5 ,globalMax: 6
the maximum subarray sum: 6
*/
